### PR TITLE
ROX-26317: Generate custom resource type for save-as function

### PR DESCRIPTION
**Pull request by c-du**
### Description

Generate a custom resource from a policy.

### User-facing documentation

- [ ] CHANGELOG is updated **OR** update is not needed
- [ ] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

### Testing and quality

- [ ] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [ ] CI results are inspected

#### Automated testing

- [ ] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [ ] modified existing tests

#### How I validated my change

change me!

### CODE REVIEW

File: central/policy/service/customresource/testdata/custom_resource.yaml [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  - cluster: prod cluster
+    namespace: stackrox
+    label:
+      key: com.docker.stack.namespace
+      value: prevent
+  severity: LOW_SEVERITY
+  enforcementActions:
+  - FAIL_BUILD_ENFORCEMENT
+  - KILL_POD_ENFORCEMENT
+  policyVersion: "1.1"
```
Comment (clickboo) [inline central/policy/service/customresource/testdata/custom_resource.yaml:47]: Just ran an experiment on my PR image which removes this - so we will need to remove this from save as cr output as well. if the incoming yaml has it, CR validation fails (k8s path, not my code)

Comment (c-du) [reply central/policy/service/customresource/testdata/custom_resource.yaml]: Removed

File: central/policy/service/customresource/custom_resource.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+type CustomResource struct {
+	APIVersion string                 `yaml:"apiVersion"`
+	Kind       string                 `yaml:"kind"`
+	Metadata   map[string]interface{} `yaml:"metadata,omitempty"`
+	Spec       *Policy                `yaml:"spec"`
+}
+
+// GenerateCustomResource generate custom resource in YAML text from a policy
+func GenerateCustomResource(policy *storage.Policy) (string, error) {
+	convertedPolicy := ConvertPolicyToCustomResource(policy)
```
Comment (clickboo) [inline central/policy/service/customresource/custom_resource.go:24]: naming suggestion for readability: policyCR

Comment (clickboo) [reply central/policy/service/customresource/custom_resource.go:24]: Where is this function `ConvertPolicyToCustomResource` defined?

File: central/policy/service/customresource/custom_resource.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	"google.golang.org/protobuf/types/known/timestamppb"
+	"gopkg.in/yaml.v3"
+)
+
+// CustomResource represents the structure for a SecurityPolicy Kubernetes Custom Resource.
+type CustomResource struct {
+	APIVersion string                 `yaml:"apiVersion"`
+	Kind       string                 `yaml:"kind"`
+	Metadata   map[string]interface{} `yaml:"metadata,omitempty"`
+	Spec       *Policy                `yaml:"spec"`
```
Comment (clickboo) [inline central/policy/service/customresource/custom_resource.go:19]: We are renaming to SecurityPolicy

File: central/policy/service/customresource/custom_resource.go (hunk 2 above)
Comment (c-du) [reply central/policy/service/customresource/custom_resource.go:24]: It is in policy.go.

File: central/policy/service/customresource/custom_resource.go (hunk 3 above)
Comment (c-du) [reply central/policy/service/customresource/custom_resource.go:19]: Its yaml representation is still "spec", so it does not matter what is named here. I see the current naming is SecurityPolicySpec.

**Review by kylape — COMMENTED** [review]
Would it be possible to hook this up to the gRPC server in this PR?  (Please correct me if I missed it)

File: central/policy/service/customresource/policy.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	}
+
+	return &PolicySection{
+		SectionName:  p.SectionName,
+		PolicyGroups: sliceutils.ConvertSlice(p.PolicyGroups, convertPolicyGroup),
+	}
+}
+
+// Policy represents storage.Policy in the Custom Resource.
+type Policy struct {
```
[the remark this answers is not in the document]
Comment (c-du) [reply central/policy/service/customresource/policy.go:111]: I create a tool that parse the tag in protobuf message. The tag looks like this:
// gotags: "crYaml:name,omitempty,stringer"
The name and omitempty are passed to yaml to generate. timestamp is telling the tool to convert this field to a string.

**Comment by c-du** [comment]
> Would it be possible to hook this up to the gRPC server in this PR? (Please correct me if I missed it)
I would rather to keep them seperate in evolving because this PR has a generator behind it. The codes in this PR was generated by the (not ready) generator and the gRPC server is in another PR https://github.com/stackrox/stackrox/pull/12415.

**Comment by kylape** [comment]
I see you've merged this PR branch into #12415.  Do you want to go ahead and merge that PR along with these changes and keep this PR open until the generator is ready to be reviewed?

**Comment by kylape** [comment]
Another question:  How feasible would it be to _also_ generate the CRD go structs using your new code generator?  We'd have to add in the kubebuilder annotations.

File: central/policy/service/customresource/policy.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (clickboo) [reply central/policy/service/customresource/policy.go:111]: Can we reuse the structs like Kyle said. We have a set of structs defined to generate the CRD as well. Check policy_types.go.

**Comment by c-du** [comment]
> I see you've merged this PR branch into #12415. Do you want to go ahead and merge that PR along with these changes and keep this PR open until the generator is ready to be reviewed?

**Comment by c-du** [comment]
> Another question: How feasible would it be to _also_ generate the CRD go structs using your new code generator? We'd have to add in the kubebuilder annotations.

Yes, that is also possible.

File: central/policy/service/customresource/policy.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (c-du) [reply central/policy/service/customresource/policy.go:111]: We cannot share the types defined in policy_types.go but like kyle said, we may use the same tool to generate them.

File: central/policy/service/customresource/custom_resource.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	// Truncate to 253 characters max, as per DNS subdomain name requirements
+	if len(name) > 253 {
+		name = name[:253]
+	}
+
+	name = strings.Trim(name, "-.")
+
+	// If the result is empty or invalid after processing, return a default value
+	if len(name) == 0 {
+		name = "rhacs.default-policy-name"
```
Comment (kylape) [inline central/policy/service/customresource/custom_resource.go:75]: Can we append a random string to ensure uniqueness?  then we could name it something like `policy-b3rd2`.

File: central/policy/service/customresource/custom_resource.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+	if t == nil {
+		return ""
+	}
+	return t.Format(time.RFC3339)
+}
+
+// toDNSSubdomainName converts a policy name to a valid DNS subdomain name.
+func toDNSSubdomainName(name string) string {
+	name = strings.ToLower(name)
+	name = strings.ReplaceAll(name, " ", ".")
```
Comment (kylape) [inline central/policy/service/customresource/custom_resource.go:58]: My impression is that dashes are used much more often than dots in k8s names.

File: central/policy/service/customresource/custom_resource_test.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+)
+
+//go:embed testdata/custom_resource.yaml
+var templateFile string
+
+func TestConvertToCR(t *testing.T) {
+	policy := getTestPolicy()
+	converted, err := GenerateCustomResource(policy)
+	require.NoError(t, err)
+	fmt.Println(converted)
```
Comment (kylape) [inline central/policy/service/customresource/custom_resource_test.go:23]: Should this be removed before merge?

File: central/policy/service/customresource/policy.go [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+		Scope: convertScope(p.Scope),
+	}
+}
+
+// Exclusion represents storage.Exclusion in the Custom Resource.
+type Exclusion struct {
+	Name       string                   `yaml:",omitempty"`
+	Deployment *Exclusion_Deployment    `yaml:",omitempty"`
+	Image      *storage.Exclusion_Image `yaml:",omitempty"`
+	Expiration string                   `yaml:",omitempty"`
```
Comment (kylape) [inline central/policy/service/customresource/policy.go:53]: do none of these fields need the name of the yaml field in the tag?

**Review by kylape — COMMENTED** [review]
Did you mean to close this PR?

File: central/policy/service/customresource/policy.go (hunk 8 above)
Comment (c-du) [reply central/policy/service/customresource/policy.go:53]: omitempty is not by default.

File: central/policy/service/customresource/custom_resource_test.go (hunk 7 above)
Comment (c-du) [reply central/policy/service/customresource/custom_resource_test.go:23]: I want to keep it for the convenience of debugging

File: central/policy/service/customresource/custom_resource.go (hunk 6 above)
Comment (c-du) [reply central/policy/service/customresource/custom_resource.go:58]: Both dot and dashes are allowed. Dot are used more often in separating the field. I am ok with both.

File: central/policy/service/customresource/custom_resource.go (hunk 5 above)
Comment (c-du) [reply central/policy/service/customresource/custom_resource.go:75]: Yes, I can do that.

**Comment by c-du** [comment]
I might accidentally close this, but thank you for reviewing it even if I do. I'm opening it now.

File: central/policy/service/customresource/policy.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kylape) [reply central/policy/service/customresource/policy.go:111]: https://issues.redhat.com/browse/ROX-26482

File: central/policy/service/customresource/custom_resource.go [hunk 9]
```diff
@@ -72,7 +74,9 @@ func toDNSSubdomainName(name string) string {
 
 	// If the result is empty or invalid after processing, return a default value
 	if len(name) == 0 {
-		name = "rhacs.default-policy-name"
+		// Get a random alphanumeric string from the first segment of a uuid
+		randAlphaNumerics := strings.Split(uuid.NewV4().String(), "-")
+		name = fmt.Sprintf("rhacs.%s", randAlphaNumerics[0])
```
Comment (kylape) [inline central/policy/service/customresource/custom_resource.go:79]: Can we use a dot instead of a dash here to be consistent with the non-default value?

**Review by kylape — APPROVED** [review]

**Pull request merged**